File: packages/native-shell-ios/Sources/SelfNativeShell/Handlers/SecureStorageHandler.swift
// SPDX-License-Identifier: BUSL-1.1

import Foundation

final class SecureStorageHandler: BridgeHandler {
    let domain: BridgeDomain = .secureStorage

    private let provider: SecureStorageProvider

    init(provider: SecureStorageProvider) {
        self.provider = provider
    }

    func handle(method: String, params: [String: Any]?) async throws -> Any? {
        let result: Any?

        switch method {
        case "get":
            guard let key = params?["key"] as? String else {
                throw BridgeHandlerError.missingParam("key")
            }
            if let value = provider.get(key: key) {
                result = ["value": value]
            } else {
                result = ["value": NSNull()]
            }

        case "set":
            guard let key = params?["key"] as? String else {
                throw BridgeHandlerError.missingParam("key")
            }
            guard let value = params?["value"] as? String else {
                throw BridgeHandlerError.missingParam("value")
            }
            try provider.set(key: key, value: value)
            result = nil

        case "remove":
            guard let key = params?["key"] as? String else {
                throw BridgeHandlerError.missingParam("key")
            }
            try provider.remove(key: key)
            result = nil

        default:
            throw BridgeHandlerError.unknownMethod(method)
        }
        return result
    }
}

File: packages/native-shell-android/src/main/kotlin/xyz/self/sdk/handlers/SecureStorageHandler.kt
// SPDX-License-Identifier: BUSL-1.1

package xyz.self.sdk.handlers

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import xyz.self.sdk.api.SecureStorageProvider
import xyz.self.sdk.bridge.BridgeDomain
import xyz.self.sdk.bridge.BridgeHandler
import xyz.self.sdk.bridge.BridgeHandlerException

class SecureStorageHandler(private val provider: SecureStorageProvider) : BridgeHandler {
    override val domain = BridgeDomain.SECURE_STORAGE

    override suspend fun handle(
        method: String,
        params: Map<String, JsonElement>,
    ): JsonElement? = when (method) {
        "get" -> get(params)
        "set" -> set(params)
        "remove" -> remove(params)
        else -> throw BridgeHandlerException("METHOD_NOT_FOUND", "Unknown secureStorage method: $method")
    }

    private fun get(params: Map<String, JsonElement>): JsonElement {
        val key = params["key"]?.jsonPrimitive?.content
            ?: throw BridgeHandlerException("MISSING_KEY", "Key parameter required")
        val value = provider.get(key)
        return buildJsonObject {
            put("value", if (value != null) JsonPrimitive(value) else JsonNull)
        }
    }

    private fun set(params: Map<String, JsonElement>): JsonElement? {
        val key = params["key"]?.jsonPrimitive?.content
            ?: throw BridgeHandlerException("MISSING_KEY", "Key parameter required")
        val value = params["value"]?.jsonPrimitive?.content
            ?: throw BridgeHandlerException("MISSING_VALUE", "Value parameter required")
        provider.set(key, value)
        return null
    }

    private fun remove(params: Map<String, JsonElement>): JsonElement? {
        val key = params["key"]?.jsonPrimitive?.content
            ?: throw BridgeHandlerException("MISSING_KEY", "Key parameter required")
        provider.remove(key)
        return null
    }
}
